routines/Random.asm

```diff
@@ -3,32 +3,42 @@
 ; Output: A = random number in the interval [0, A]
  
 ?main:
-    PHX : PHP
-    SEP #$30
-    PHA
-    JSL $01ACF9|!BankB
-    LDA $148D|!Base2
-    PLX
-    CPX #$FF
-    BEQ ?.end
- 
-?.normal
-    INX
- 
-    if !SA1 == 0
-        STA $4202               ; Write first multiplicand.
-        STX $4203               ; Write second multiplicand.
-        NOP #4                  ; Wait 8 cycles.
-        LDA $4217               ; Read multiplication product (high byte).
+    phx : php
+    sep #$30
+
+    pha
+    inc
+    and $01,s
+    beq ?.powOf2                        ; max of the form 2^n - 1, so we can just use a bitwise and instead of a divide
+
+?.divide:
+    if !sa1 == 0
+        jsl $01ACF9|!bank               ; first byte in both A and $148C, second in $148D
+        sta $4204                       ; dividend, low byte
+        lda $148D|!addr : sta $4205     ; dividend, high byte
+        pla : inc
+        sta $4206                       ; divisor
+        nop #8                          ; wait 16 cycles

```

Comment:
16 cycles is quite the speed penalty here, is there any way to lessen the impact here?

Reply:
MFG said that 7 nops should be enough (similar reasoning to the sa-1 version).  I don't really have a way to test real hardware or things like fpga emulators, so I usually just err on the side of caution as far as that goes.  Otherwise, the existing routine was just essentially using a clever trick to do a multiply in order to do an 8-bit division, saving some cycles.  But only using 8 bits of randomness was the whole cause of the modulo bias in the first place, and I don't think there's a way to use the same trick for 16 bits.

Reply:
Is the removed modulo bias worth the speed penalty in this case?

Reply:
I think so, at least.  Especially when it's an extra 8 cycles on top of the call to the RNG routine itself along with some other overhead, it's unlikely to be particularly noticeable.  And like I was saying earlier, for someone who wants a number in the 0-2 range, it's probably not a big deal, but for someone looking to get a random number in the 0-99 range, doing so by taking the remainder of an 8-bit random value mod 100 is going to give really unacceptable results.